Create NestedJson adapter.

kayhide opened the pull request:
I needed an adapter to serialize models with multi-depth relations for my project.
So I wrote it.

I think this might be helpful for some people.

joaomdmoura commented:
wow! Didn't reviewed it yet, but this is awesome, I'm definitely interest on this. I'll check tonight probably.

kayhide commented:
Hi,
I have checked the discussion on #921.

This adapter raises an error if nesting is too deep, typically in case of circular relations.
I think this is a desirable behavior for many cases.

If you want to change the limit depth, you can do it just giving an argument `limit_depth: xx` to the adapter, or setting default value `NestedJson.default_limit_depth = xx`.

bf4 commented:
Possibly related:

- https://github.com/rails-api/active_model_serializers/issues/211
- https://github.com/rails-api/active_model_serializers/pull/789

bf4 commented on lib/active_model/serializer/adapter/nested_json.rb:
```diff
@@ -0,0 +1,57 @@
+        cattr_accessor :default_limit_depth
+        @@default_limit_depth = 5
+
+        def serializable_hash options = {}
```
Yikes, it's a pretty big smell in the Json serializer that we're subclassing it but completely overriding this method, where everything important happens

also, might this belong in the Adapter superclass? It looks on first inspection like it would apply to the JsonApi adapter as well, no?

kayhide replied:
That is for `#fragment_cache`.

The way we cache fragmentation is exactly the same to `Adapter::Json` ( like `Adapter::FlattenJson` as well).
And `Adapter::JsonApi` caches differently.

bf4 commented on test/fixtures/poro.rb:
```diff
@@ -266,3 +266,18 @@ def json_key
+AuthorNestedSerializer = Class.new(ActiveModel::Serializer) do
+  attributes :id, :name
+  has_many :posts, serializer: PostNestedSerializer
+end
```
@joaomdmoura I think we should start following some convention for associating poros with tests. I think it's becoming append-only...

kayhide replied:
I have a suggestion.
In my projects, I am doing like this way:

```ruby
class Author < ActiveModel::Serializer
  attributes :id, :name

  class ForShow < self
    has_many :posts
  end
end

class Post < ActiveModel::Serializer
  attributes :id, :title, :body

  class ForShow < self
    has_many :comments
  end
end
```

and so on.

I define default serializers not to have any  relations.
And for specific cases, like showing details, define serializers as nested classes.

When I use them in a controller, I write like:
```ruby
def show
  ...
  render json: @post, serializer: PostSerializer::ForShow
end
```

Like this way, I can easily avoid unexpected circular relations.
Also scoping with default serializer classes make it clear which serializer is for which model.

bf4 replied:
works for me

bf4 commented:
Thanks!

joaomdmoura commented:
Hey, sorry for taking so long, was really busy, now I stopped to check it our for real.
First of all tks @kayhide and @bf4 for your hard work. I really appreciate this. 

I know that you needed multi-depth relations and make your own adapter was the fastest way, but IMO the best scenario would be to find a way to add it to json + flatten-json existing adapters. What are your thoughts about it?  :smile:

kayhide commented:
@joaomdmoura I agree with you.

OK, I will add these functionalities on this adaptor.

- taking care of root
- adding an option to trim off deep associations not to raise an error

After covering these, I think we can replace the existing json, flatten-json adapaters.

kayhide commented:
Hi, I updated the adapter and replaced existing Json/FlattenJson adapter.

This adapter accepts options:
`limit_depth` limits the depth of associations.

`check_depth_strategy` defines the behavior when nesting is too deep.
if `:fail` is set, it raises an error.
if `:trim` is set, it ignores deeper associations.

Setting `limit_depth = 1` and `check_depth_strategy = :trim` makes its behavior just the same to the original Json/FlattenJson adapter.

And default values are available setting like:
```ruby
Adapter::Json.default_limit_depth = 5
Adapter::Json.default_check_depth_strategy = :fail
```

For now, I set default values to follow original adapters behavior, and made them to pass all the existing tests untouched.
But I think people want it to handle nested associations to some depth as a default.

To do so, we need to fix the default values and rewrite the relevant tests.

bryanrite commented:
I really like this approach, defining all appropriate associations in each serializer and then being able to limit the depth in the controller is perfect.

bryanrite commented:
One bug I noticed will trying out this branch, is that attributes with the name `name` cause an error now. For example, if the `Comment` model had a "name" attribute and you were serializing it like: 

```ruby
def Api::V1::CommentSerializer < ActiveModel::Serializer
  attribute :id
  attribute :name
end
```

you get:

```ruby
undefined method `name' for #<Api::V1::CommentSerializer:0x007fdec7a23408>`
```

This is fixed by defining a method:

```ruby
def name
  object.name
end
```

This is the only attribute it happens on that I have found so far.

bryanrite commented:
And +1 for supporting a `limit_depth` of 0.

kayhide commented:
@bryanrite hi, thanks for your support :)

> attributes with the name name cause an error 

You can try `attributes` not `attribute`.
It's plural not singular.
Those are different methods.

As I read the code, singular version is for the case when the key and the accessor method are not the same.

beauby commented:
The error is related to #1094.

kayhide commented:
Updated to support limitless depth.
Giving `check_depth_strategy = :pass` makes it go deep without any limit.
In this case the `limit_depth` value is ignored.

kayhide commented:
@beauby ah, right. thanks for the comment.

bf4 commented on lib/active_model/serializer/adapter/json.rb:
```diff
@@ -4,49 +4,86 @@ module ActiveModel
-        def fragment_cache(cached_hash, non_cached_hash)
-          Json::FragmentCache.new().fragment_cache(cached_hash, non_cached_hash)
+        def too_deep? depth
+          if depth > @limit_depth
```
nice!

NullVoxPopuli commented:
closed in favor of #1127 and #1158 
deep relationships are now managed via the include option passed in to the adapter

example:

`render json: @posts, include: [author, comments: author]`

or

`render json: @posts, include: 'author,comments.author'`

beauby commented:
Moreover, the "depth strategy" is available via wildcards:
```ruby
render json: @posts, include: '*.*.*'
```
This would include all nested relationships 3 levels down.
